File: src/robot_driver/robot_driver/TurnRightActionServer.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from bot_behavior_interfaces.action import TurnRight  # Import the TurnRight action
from geometry_msgs.msg import Twist

class TurnRightActionServer(Node):
    def __init__(self):
        super().__init__('turn_right_action_server')
        
        # Declare and set the turning speed 
        self.turn_speed = -0.5  

        self._action_server = ActionServer(
            self,
            TurnRight,
            'turn_right',
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback)
        self.publisher_ = self.create_publisher(Twist, 'cmd_vel', 10)
        self.timer = None

    def goal_callback(self, goal_request):
        self.get_logger().info('Received request to turn right.')
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        self.get_logger().info('Received cancel request. Stopping the robot.')
        self.stop_robot()
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        self.get_logger().info(f'Executing turn right with speed {self.turn_speed}...')
        
        twist_msg = Twist()
        twist_msg.angular.z = self.turn_speed  # Use the class attribute for angular speed

        # Create a timer to publish twist messages periodically
        self.timer = self.create_timer(0.1, self.publish_twist_msg)
        self.timer_msg = twist_msg

        while not goal_handle.is_cancel_requested:
            rclpy.spin_once(self, timeout_sec=0.1)

        if goal_handle.is_cancel_requested:
            self.get_logger().info('Goal canceled. Stopping the robot.')
            self.stop_robot()
            goal_handle.canceled()
            return TurnRight.Result()

    def publish_twist_msg(self):
        if self.timer_msg:
            self.publisher_.publish(self.timer_msg)

    def stop_robot(self):
        # Publish zero velocity to stop the robot
        twist_msg = Twist()
        twist_msg.linear.x = 0.0
        twist_msg.angular.z = 0.0
        self.publisher_.publish(twist_msg)
        if self.timer:
            self.timer.cancel()

def main(args=None):
    rclpy.init(args=args)
    turn_right_action_server = TurnRightActionServer()
    rclpy.spin(turn_right_action_server)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/robot_driver_cpp/src/robot_driver_cpp/turn_right_action_server.cpp
#include "robot_driver_cpp/turn_right_action_server.hpp"

using namespace std::chrono_literals;

TurnRightActionServer::TurnRightActionServer() : Node("turn_right_action_server")
{
    this->declare_parameter<double>("turn_speed", 0.2);
    this->get_parameter("turn_speed", turn_speed_);

    action_server_ = rclcpp_action::create_server<TurnRight>(
        this,
        "turn_right",
        std::bind(&TurnRightActionServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
        std::bind(&TurnRightActionServer::handle_cancel, this, std::placeholders::_1),
        std::bind(&TurnRightActionServer::handle_accepted, this, std::placeholders::_1)
    );

    publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
}

rclcpp_action::GoalResponse TurnRightActionServer::handle_goal(
    const rclcpp_action::GoalUUID &uuid,
    std::shared_ptr<const TurnRight::Goal> goal
)
{
    RCLCPP_INFO(this->get_logger(), "Received request to turn right.");
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse TurnRightActionServer::handle_cancel(
    const std::shared_ptr<GoalHandleTurnRight> goal_handle
)
{
    RCLCPP_INFO(this->get_logger(), "Received cancel request. Stopping the robot.");
    stop_robot();
    return rclcpp_action::CancelResponse::ACCEPT;
}

void TurnRightActionServer::handle_accepted(const std::shared_ptr<GoalHandleTurnRight> goal_handle)
{
    std::thread{std::bind(&TurnRightActionServer::execute, this, std::placeholders::_1), goal_handle}.detach();
}

void TurnRightActionServer::execute(const std::shared_ptr<GoalHandleTurnRight> goal_handle)
{
    RCLCPP_INFO(this->get_logger(), "Executing turn right with speed %f...", turn_speed_);
    
    twist_msg_.angular.z = -turn_speed_;
    timer_ = this->create_wall_timer(
        100ms, [this]() { publish_twist_msg(); }
    );

    while (rclcpp::ok() && !goal_handle->is_canceling()) {
        std::this_thread::sleep_for(100ms);
    }

    if (goal_handle->is_canceling()) {
        RCLCPP_INFO(this->get_logger(), "Goal canceled. Stopping the robot.");
        stop_robot();
        goal_handle->canceled(std::make_shared<TurnRight::Result>());
    } else {
        goal_handle->succeed(std::make_shared<TurnRight::Result>());
    }
}

void TurnRightActionServer::publish_twist_msg()
{
    if (publisher_) {
        publisher_->publish(twist_msg_);
    }
}

void TurnRightActionServer::stop_robot()
{
    twist_msg_.linear.x = 0.0;
    twist_msg_.angular.z = 0.0;
    publisher_->publish(twist_msg_);
    if (timer_) {
        timer_->cancel();
    }
}
